### PR TITLE
[export] enumerate unsupported sympy.Functions

### PR DESCRIPTION
**Pull request by pianpwk**
There's 2 concepts of unsupported sympy.Functions in symbolic_shapes:
1) unsupported by the export solver, meaning the solver doesn't know how to provide useful fixes for those functions
2) unsupported by the sympy interpreter - meaning we can't reify them into FX nodes because the functions aren't present in PythonReferenceAnalysis

This splits the current call into a call for each version, with the Export solver the only user of 1). For 1), we enumerate the functions in _sympy/functions.py, and subtract the functions we know we can support. For 2) there's only 3 functions we've seen pop up in test cases.

Differential Revision: D61677956

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

@diff-train-skip-merge
// reverted internally

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -1741,6 +1751,17 @@ def floor_div_handler(*args):
             expr = expr.replace(FloorDiv, floor_div_handler)
         return expr
 
+    def _enumerate_sympy_functions(self):
+        module = torch.utils._sympy.functions
+        all_functions = {getattr(module, func) for func in module.__all__}
+        self._unsupported_sympy_functions = all_functions.difference(self._supported_sympy_functions)
+
+    def _has_unsupported_sympy_function(self, expr) -> bool:
```
Comment (pianpwk) [inline torch/fx/experimental/symbolic_shapes.py:1759]: @avikchaudhuri sorry, adding this back. I know you had this in the first place, I just didn't fully understand it at the time.

File: torch/fx/experimental/symbolic_shapes.py [hunk 2]
```diff
@@ -1675,6 +1676,15 @@ def __init__(
         # symbols that are marked dynamic
         self._marked_dynamic = marked_dynamic
 
+        # track unsupported sympy functions
+        self._supported_sympy_functions: Set[sympy.Function] = {
+            Mod,
+            PythonMod,
+            FloorDiv,
+        }
+        self._unsupported_sympy_functions: Set[sympy.Function] = {}
```
Comment (avikchaudhuri) [inline torch/fx/experimental/symbolic_shapes.py:1685]: Adjust the initializations and comments here, `_supported_sympy_functions` is a fixed list and `_unsupported_sympy_functions` will be immediately populated by enumeration? So maybe no need to initialize `_unsupported_sympy_functions` here or just inline the enumeration.

File: torch/fx/experimental/symbolic_shapes.py [hunk 3]
```diff
@@ -1741,6 +1751,17 @@ def floor_div_handler(*args):
             expr = expr.replace(FloorDiv, floor_div_handler)
         return expr
 
+    def _enumerate_sympy_functions(self):
+        module = torch.utils._sympy.functions
+        all_functions = {getattr(module, func) for func in module.__all__}
```
Comment (avikchaudhuri) [inline torch/fx/experimental/symbolic_shapes.py:1756]: One issue here is that `__all__` does not enumerate everything in that module (e.g., `FloorToInt`), please check. Might have to get dirty here.

**Review by avikchaudhuri — APPROVED** [review]
Stamping to unblock, see comments.

**Comment by malfet** [comment]
Somehow this broke export tests on MacOS, see https://hud.pytorch.org/hud/pytorch/pytorch/4df10a634098135f80dfb1fa9f1be56461f2deed/1?per_page=50&name_filter=macos and https://github.com/pytorch/pytorch/actions/runs/10568420165/job/29279626601

**Pull request closed without merging**